pkg/instrumentation/python.go

```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package instrumentation
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/open-telemetry/opentelemetry-operator/apis/instrumentation/v1alpha1"
+)
+
+const (
+	envPythonPath         = "PYTHONPATH"
+	envOtelTracesExporter = "OTEL_TRACES_EXPORTER"
+	pythonPathPrefix      = "/otel-auto-instrumentation/opentelemetry/instrumentation/auto_instrumentation"
+	pythonPathSuffix      = "/otel-auto-instrumentation"
+)
+
+func injectPythonSDK(logger logr.Logger, pythonSpec v1alpha1.PythonSpec, pod corev1.Pod) corev1.Pod {
+	// caller checks if there is at least one container
+	container := &pod.Spec.Containers[0]
+	idx := getIndexOfEnv(container.Env, envPythonPath)
+	if idx == -1 {
+		container.Env = append(container.Env, corev1.EnvVar{
+			Name:  envPythonPath,
+			Value: fmt.Sprintf("%s:%s", pythonPathPrefix, pythonPathSuffix),
+		})
+	} else if idx > -1 {
+		if container.Env[idx].ValueFrom != nil {
+			// TODO add to status object or submit it as an event
+			logger.Info("Skipping Python SDK injection, the container defines PYTHONPATH env var value via ValueFrom", "container", container.Name)
+			return pod
+		}
+		container.Env[idx].Value = fmt.Sprintf("%s:%s:%s", pythonPathPrefix, container.Env[idx].Value, pythonPathSuffix)
+	}
+
+	// Set OTEL_TRACES_EXPORTER to HTTP exporter if not set by user because it is what our autoinstrumentation supports.
```

Comment:
python does not support OTLP GRPC?

Reply:
It does but uses a native extension wrapping gRPC-C. This means the extension can only be loaded on the same python version as when pip install was invoked.

HTTP transport is pure python so avoids causing potential version problems.

Reply:
By the way related comment here

https://github.com/open-telemetry/opentelemetry-operator/pull/524#discussion_r746518160